types/src/v0/impls/state.rs

```diff
@@ -306,6 +316,21 @@ pub fn validate_proposal(
         });
     }
 
+    // Check if timestamp is increasing (with some tolerance).
+    if proposal.timestamp() <= parent_header.timestamp() - 2 {
+        return Err(ProposalValidationError::InvalidTimestamp {
+            parent_timestamp: parent_header.timestamp(),
+            proposal_timestamp: proposal.timestamp(),
+        });
+    }
+
+    if proposal.l1_head() <= parent_header.l1_head() {
```

Comment:
We should also wait for the `l1_head` to actually exist (ie wait for L1 block number >= `proposal.l1_head()`) as we do for the finalized block. Otherwise a malicious proposer can set `l1_head` arbitrarily far in the future, and it ceases to be a useful field

Reply:
dc197d664cf7cc2368500b08ed584a52b4c9398f